[SPARK-29748][PYTHON][SQL] Remove Row field sorting in PySpark for version 3.6+

BryanCutler opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Removing the sorting of PySpark SQL Row fields that were previously sorted by name alphabetically for Python versions 3.6 and above. Field order will now match that as entered. Rows will be used like tuples and are applied to schema by position. For Python versions < 3.6, the order of kwargs is not guaranteed and therefore will be sorted automatically as in previous versions of Spark.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
This caused inconsistent behavior in that local Rows could be applied to a schema by matching names, but once serialized the Row could only be used by position and the fields were possibly in a different order.

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
Yes, Row fields are no longer sorted alphabetically but will be in the order entered. For Python < 3.6 `kwargs` can not guarantee the order as entered, so `Row`s will be automatically sorted. 

An environment variable "PYSPARK_ROW_FIELD_SORTING_ENABLED" can be set that will override construction of `Row` to maintain compatibility with Spark 2.x. 


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Existing tests are run with PYSPARK_ROW_FIELD_SORTING_ENABLED=true and added new test with unsorted fields for Python 3.6+

BryanCutler committed:
Remove Row field sorting and add LegacyRow with env var

BryanCutler committed:
Start fixing tests

BryanCutler commented:
WIP still need to do:

- [ ] Add note to migration guide
- [ ] Check docs
- [ ] Fix remaining test failures
- [ ] Add new tests

SparkQA commented:
**[Test build #113676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113676/testReport)** for PR 26496 at commit [`ecd8f83`](https://github.com/apache/spark/commit/ecd8f83405ebabba6408a1e031147fce9832cf86).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18560/
Test FAILed.

SparkQA commented:
**[Test build #113676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113676/testReport)** for PR 26496 at commit [`ecd8f83`](https://github.com/apache/spark/commit/ecd8f83405ebabba6408a1e031147fce9832cf86).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113676/
Test FAILed.

BryanCutler committed:
Fix dataframe doctests

BryanCutler committed:
Fix sql.types doctests

BryanCutler committed:
Fix sql.functions doctests

BryanCutler committed:
Made _LegacyRow private, added pydoc

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/18691/
Test PASSed.

SparkQA commented:
**[Test build #113819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113819/testReport)** for PR 26496 at commit [`b85c3a6`](https://github.com/apache/spark/commit/b85c3a606a2e8e77f59d31777500fc6acebc828a).

SparkQA commented:
**[Test build #113819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113819/testReport)** for PR 26496 at commit [`b85c3a6`](https://github.com/apache/spark/commit/b85c3a606a2e8e77f59d31777500fc6acebc828a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class _LegacyRow(Row):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113819/
Test FAILed.

BryanCutler commented:
ping @HyukjinKwon @viirya for thoughts on the proposed implementation. For testing of Python 2.7, I think we will have to set the env variable to use LegacyRow by default, otherwise there is lots of code to change.

HyukjinKwon reviewed (commented):
Sorry for my late response. +1 from me.

HyukjinKwon commented on python/pyspark/sql/types.py:
```diff
@@ -1463,32 +1474,43 @@ class Row(tuple):
+
         if kwargs:
             # create row objects
-            names = sorted(kwargs.keys())
```
Actually, after a second thought, why don't we just have an env to switch on and off the sorting, and disable it in Spark 3.1? I think it will cause less changes I suspect.

BryanCutler replied:
Yeah, we could do that but that doesn't solve the problem of the `__from_dict__` flag that is not needed if there is no sorting. That flag isn't serialized which causes different behavior when serialized.

BryanCutler replied:
Hmm, actually it looks like it could be possible to only add the `__from_dict__` flag if sorting is enabled too. I can give that a try and see if it works, wdyt?

BryanCutler committed:
Env var now controls sorting only, removed LegacyRow and OrderedDict as arg

BryanCutler committed:
Revert all test fixes

BryanCutler committed:
Set PYSPARK_ROW_FIELD_SORTING_ENABLED to 'true' for run_tests

BryanCutler committed:
Added test with sorting disabled

SparkQA commented:
**[Test build #114803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114803/testReport)** for PR 26496 at commit [`832545b`](https://github.com/apache/spark/commit/832545b978bff69155adb4e1c5026f106333d4ea).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19625/
Test PASSed.

BryanCutler commented:
Ok, I changed to @HyukjinKwon suggestion of using the env var to control sorting only, and also creation of the `__from_dict__` flag. I also change to error for Python < 3.6 to generate a warning and fallback to sorting fields and removed the option for using an OrderedDict.

I think this is the most gentle approach to introducing new behavior (of not sorting) and still allowing users to create legacy Rows that are sorted, if needed. It will also be a fairly clean removal when Python < 3.6 is deprecated. I did have to set the env var by default for tests, but I introduced a new tests that makes an unsorted Row.

Let me know if there is any issue with the current implementation, and I will add to the migration guide as soon as I can.

SparkQA commented:
**[Test build #114803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114803/testReport)** for PR 26496 at commit [`832545b`](https://github.com/apache/spark/commit/832545b978bff69155adb4e1c5026f106333d4ea).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114803/
Test FAILed.

BryanCutler committed:
Move flag to static member of Row, fix test to skip for Python < 3.6

SparkQA commented:
**[Test build #114804 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114804/testReport)** for PR 26496 at commit [`8a8f2aa`](https://github.com/apache/spark/commit/8a8f2aa7d94a34f48c38a84672e2bf60d8f44946).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19626/
Test PASSed.

SparkQA commented:
**[Test build #114804 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114804/testReport)** for PR 26496 at commit [`8a8f2aa`](https://github.com/apache/spark/commit/8a8f2aa7d94a34f48c38a84672e2bf60d8f44946).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon commented on python/pyspark/sql/tests/test_types.py:
```diff
@@ -968,6 +968,19 @@ def __init__(self, **kwargs):
                 _make_type_verifier(data_type, nullable=False)(obj)
 
+    @unittest.skipIf(sys.version_info[:2] < (3, 6), "Create Row without sorting fields")
+    def test_Row_without_field_sorting(self):
```
no big deal but can we `test_Row_without_field_sorting` -> `test_row_without_field_sorting`? Strictly it follows pep8 I guess (and I personally don't like such names in the current codebase ... )

viirya commented on python/pyspark/sql/types.py:
```diff
@@ -1432,10 +1434,20 @@ class Row(tuple):
+
+    NOTE: As of Spark 3.0.0, the Row field names are no longer sorted
+    alphabetically. To enable field sorting to create Rows compatible with
+    Spark 2.x, set the environment variable "PYSPARK_ROW_FIELD_SORTING_ENABLED"
```
I'm curious when this compatibility will be matter?  If using Python >= 3.6 at Spark 3.0.0, do users need this compatibility?

BryanCutler replied:
Yeah, good question. I think it's possible for even users that have Python 3.6 to have code that relied on the field names being sorted and this will break their existing code. So they still might need to set the env var until the existing code could be updated. Not sure how likely this scenario is...

BryanCutler committed:
Fixed wording and grammar

SparkQA commented:
**[Test build #114814 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114814/testReport)** for PR 26496 at commit [`d2f0bed`](https://github.com/apache/spark/commit/d2f0bedfe395ec86afe737a1c041b15649768bf9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19637/
Test PASSed.

SparkQA commented:
**[Test build #114814 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114814/testReport)** for PR 26496 at commit [`d2f0bed`](https://github.com/apache/spark/commit/d2f0bedfe395ec86afe737a1c041b15649768bf9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114814/
Test PASSed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20580/
Test PASSed.

SparkQA commented:
**[Test build #115787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115787/testReport)** for PR 26496 at commit [`d2f0bed`](https://github.com/apache/spark/commit/d2f0bedfe395ec86afe737a1c041b15649768bf9).

SparkQA commented:
**[Test build #115787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115787/testReport)** for PR 26496 at commit [`d2f0bed`](https://github.com/apache/spark/commit/d2f0bedfe395ec86afe737a1c041b15649768bf9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115787/
Test PASSed.

BryanCutler committed:
Added note in migration guide

SparkQA commented:
**[Test build #116194 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116194/testReport)** for PR 26496 at commit [`3a69539`](https://github.com/apache/spark/commit/3a69539d15cbbaee40876832066a1a49b077e8d4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20987/
Test PASSed.

BryanCutler commented:
Apologies for the delay, but I've updated with a note in the migration guide and rebased, and removed the WIP. Please take another look @HyukjinKwon @viirya , thanks!

SparkQA commented:
**[Test build #116194 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116194/testReport)** for PR 26496 at commit [`3a69539`](https://github.com/apache/spark/commit/3a69539d15cbbaee40876832066a1a49b077e8d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116194/
Test PASSed.

HyukjinKwon reviewed (commented):
Looks okay though I haven;t taken a super close look. Should be good to go @BryanCutler if you feel sure. I will take a closer look otherwise in some days.

BryanCutler commented:
Thanks @HyukjinKwon , I'll give it a few days for any more comments. I'm not crazy about python <3.6 defaulting to the old behavior, but soon we will drop those and it won't be an issue anymore.

BryanCutler committed:
Clarified wording in doc and added a note about example output

SparkQA commented:
**[Test build #116430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116430/testReport)** for PR 26496 at commit [`14e691d`](https://github.com/apache/spark/commit/14e691d3fba4dd416d88f84d0cb5e7489cd69aa3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21218/
Test PASSed.

SparkQA commented:
**[Test build #116430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116430/testReport)** for PR 26496 at commit [`14e691d`](https://github.com/apache/spark/commit/14e691d3fba4dd416d88f84d0cb5e7489cd69aa3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116430/
Test PASSed.

BryanCutler commented:
merged to master, thanks @HyukjinKwon and @viirya

zsxwing commented on docs/pyspark-migration-guide.md:
```diff
@@ -87,6 +87,8 @@ Please refer [Migration Guide: SQL, Datasets and DataFrame](sql-migration-guide.
   - Since Spark 3.0, `Column.getItem` is fixed such that it does not call `Column.apply`. Consequently, if `Column` is used as an argument to `getItem`, the indexing operator should be used.
     For example, `map_col.getItem(col('id'))` should be replaced with `map_col[col('id')]`.
 
+  - As of Spark 3.0 `Row` field names are no longer sorted alphabetically when constructing with named arguments for Python versions 3.6 and above, and the order of fields will match that as entered. To enable sorted fields by default, as in Spark 2.4, set the environment variable `PYSPARK_ROW_FIELD_SORTING_ENABLED` to "true". For Python versions less than 3.6, the field names will be sorted alph ...
```
nit: Could we mention that this must be set for all processes? For example, `set the environment variable `PYSPARK_ROW_FIELD_SORTING_ENABLED` to "true" for **executors and driver**. This env must be consistent on all executors and driver. Any inconsistency may cause failures or incorrect answers `

HyukjinKwon replied:
+1. Let me fix it.
